modules/node-shared/src/main/scala/io/constellationnetwork/node/shared/domain/swap/block/AllowSpendBlockAcceptanceLogic.scala

```diff
@@ -0,0 +1,184 @@
+                          )
+                        ).asRight
+                      else if (txChain.head.parent.hash =!= lastTxRef.hash) // ordinals are equal
+                        RejectedAllowSpend(
+                          headTxChainRef,
+                          ParentHashNotEqLastTxHash(txChain.head.parent.hash, lastTxRef.hash)
+                        ).asLeft
+                      else // hashes and ordinals are equal
+                        (lastTxRefsUpdate.updated(address, lastTxChainRef), maybeAwaitingBlock).asRight
+
+                  } yield result
+
+                EitherT(rejectionOrUpdate)
+            }
+          }
+          .leftWiden[AllowSpendBlockNotAcceptedReason]
+          .flatMap {
+            case (update, maybeAwaitReason) =>
+              maybeAwaitReason
+                .widen[AllowSpendBlockNotAcceptedReason]
+                .toLeft(update)
+                .toEitherT[F]
+          }
+          .map { lastTxRefsUpdate =>
+            contextUpdate.copy(lastTxRefs = lastTxRefsUpdate)
+          }
+
+      private def processBalances(
+        block: Signed[AllowSpendBlock],
+        context: AllowSpendBlockAcceptanceContext[F],
+        contextUpdate: AllowSpendBlockAcceptanceContextUpdate
+      ): EitherT[F, AllowSpendBlockNotAcceptedReason, AllowSpendBlockAcceptanceContextUpdate] = {
+        val minusFn: Amount => Balance => Either[BalanceArithmeticError, Balance] = a => _.minus(a)
+        val plusFn: Amount => Balance => Either[BalanceArithmeticError, Balance] = a => _.plus(a)
+
+        val sortedTxs = block.transactions.toNonEmptyList
+        val minusAmountOps = sortedTxs.groupMap(_.source)(tx => minusFn(tx.amount))
+        val minusFeeOps = sortedTxs.groupMap(_.source)(tx => minusFn(tx.fee))
+        val plusAmountOps = sortedTxs.groupMap(_.destination)(tx => plusFn(tx.amount))
+
+        val allOps = minusAmountOps |+| minusFeeOps |+| plusAmountOps
+
+        allOps
+          .foldLeft(contextUpdate.balances.asRight[AddressBalanceOutOfRange].toEitherT[F]) { (acc, addressAndOps) =>
+            acc.flatMap { balancesUpdate =>
+              val (address, ops) = addressAndOps
+
+              EitherT(
+                balancesUpdate
+                  .get(address)
+                  .toOptionT[F]
+                  .orElseF(context.getBalance(address))
+                  .getOrElse(Balance.empty)
+                  .map { balance =>
+                    ops
+                      .foldLeft(balance.asRight[BalanceArithmeticError]) { (acc, op) =>
+                        acc.flatMap(op)
+                      }
+                      .leftMap(AddressBalanceOutOfRange(address, _))
+                      .map(balancesUpdate.updated(address, _))
+                  }
+              )
+            }
+          }
+          .leftWiden[AllowSpendBlockNotAcceptedReason]
+          .map { balancesUpdate =>
+            contextUpdate.copy(balances = balancesUpdate)
+          }
+      }
```

Comment:
@marcinwadon should we change the balances during the allow spend block acceptance? I think that we should do this after SpendTransaction, right? I could be wrong. I am sorry if I'm saying wrongly

Reply:
Yes, we should deduct the balance for the allow spend amount, to lock tokens. When SpendTransaction gets accepted, we remove allow spend and move back unused tokens to origin account. When AllowSpend expires, we need to move back tokens to the origin account.